#162948838 Implement Email verification upon registration

d-kahara opened the pull request:
**What does this PR do?**

Upon successful registration, a user should receive a verification link in the email account they provide. 

**Description of tasks to be completed?**

- Upon successful registration, the user should receive a verification link in their email
- Write tests for email verification and refactor failing tests after feature implementation
- Create view for updating user details after verification

**Export the following into your .env**
``` 
export EMAIL_HOST_USER="gmail_user_name"

export EMAIL_HOST_PASSWORD="gmail_password" 

```

How to run tests

` python3 manage.py test `

**Relevant Pivotal Tracker Stories**
[#162948838](https://www.pivotaltracker.com/story/show/162948838)

**Screenshots**
![image](https://user-images.githubusercontent.com/27801956/51320364-9cc00500-1a70-11e9-94c3-c378976e14a9.png)

d-kahara committed:
feat(email verification): Implement Email verification upon registration

-Configure gmail settings
-Update user model
-write tests for new functionality and refactor failing tests
-Update serializer and views to allow for email sending
-Update url for new route

[Delivers #162948838]

d-kahara committed:
(fix):Refactor config variables

-Refactor environment variable name to be standard across the app
-Change generic view for Verify Url View

[Fixes #162948838]

MaggieChege commented on authors/apps/authentication/views.py:
```diff
@@ -47,7 +47,7 @@ def post(self, request):
         return Response(serializer.data, status=status.HTTP_201_CREATED)
 
-class VerifyAPIView(generics.UpdateAPIView):
+class VerifyAPIView(generics.ListCreateAPIView):
```
There should be two blank lines before the beginning of this class.

kwameasiago reviewed (approved):
Looks good to me

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -110,6 +107,12 @@ def validate(self, data):
             )
 
+        #If user has not verified their email account an exception will be raised
+        if not user.is_verified:
```
Let's allow user to login even if they are not verified

d-kahara replied:
Done

ignatiusukwuoma commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,57 @@
+        )
+
+        # Test that one message has been sent.
+        self.assertEqual(len(mail.outbox), 1)
```
Before you send the email assert that mail.outbox is 0

d-kahara replied:
Done

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -1,14 +1,21 @@
 from rest_framework.views import APIView
 from rest_framework.renderers import BrowsableAPIRenderer
 from .renderers import UserJSONRenderer
+from authors.settings import SECRET_KEY, EMAIL_HOST_USER
```
import settings from django.conf

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -24,10 +31,43 @@ def post(self, request):
+        token = jwt.decode(
+            self.kwargs["token"],SECRET_KEY, algorithm='HS256')
+        queryset = User.objects.filter(email=token['email'])
+        User.objects.filter(email=token['email']).update(is_verified=True)
```
since we are expecting only one User, why don't we use `.get`

d-kahara replied:
Done

ignatiusukwuoma commented:
Let's be consistent and descriptive with file names... `email_verify_account` instead of `email-verify`

ignatiusukwuoma commented:
Please delete email-verify template

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
+    ResetSerializerEmail, ResetSerializerPassword
 )
-
+from authors.apps.core.permissions import IsAuthorOrReadOnly
```
This import is not used in this file

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
     serializer_class = RegistrationSerializer
 
+    #Method to Send emails
+    def send_link(self,email):
```
a space after `self,`. Also, this will do better as a utility function. Create a utils.py file and make this a function. Import to this file and call when needed

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
+    #Method to Send emails
+    def send_link(self,email):
+        payload = {'email':  email,"iat": datetime.now(),
+                    "exp": datetime.utcnow() + timedelta(minutes=5)}
```
Excess space after email key, missing whitespace after `email,`, put each key on its own line to aid readability

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
+        payload = {'email':  email,"iat": datetime.now(),
+                    "exp": datetime.utcnow() + timedelta(minutes=5)}
+        token = jwt.encode(payload,settings.SECRET_KEY, algorithm='HS256').decode()
+        from_email, to_email =settings.EMAIL_HOST_USER , [email]
```
Hmm... please run this file through pycodestyle. Too many issues.

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
+            '/api/users/verify/{}/'.format(token)
+
+        message = render_to_string(
+            'email-verify.html', {
```
Thought you replaced this template.

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -6,16 +12,44 @@
+        send_mail(subject,'',from_email,
+                    [to_email, ], html_message=message, fail_silently=False)
+
+        return Response({"Email verification link sent successfully"},
```
If you make this a utility function, you won't need to return Response at the end since the view's you'll call it in will return a Response

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -24,10 +58,47 @@ def post(self, request):
+                message = {
+                    "message": "Your account is already verified.",
+                }
+                return Response(message, status=status.HTTP_400_BAD_REQUEST)
```
I think a 403 is more appropriate

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -24,10 +58,47 @@ def post(self, request):
         return Response(serializer.data, status=status.HTTP_201_CREATED)
 
+class VerifyAPIView(generics.CreateAPIView):
+    serializer_class = UserSerializer
```
Do you need this Serializer? Doesn't look like you are serializing anything

d-kahara replied:
If I omit the serializer, I get an assertion error, I suppose when one is using generic views, you are required to have a serializer...not sure about this though

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -24,10 +58,47 @@ def post(self, request):
+        try:
+            email = jwt.decode(token, settings.SECRET_KEY)['email']
+            user = User.objects.get(email=email)
+            if user.is_verified == True:
```
This is the same as `if user.is_verified:`

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -24,10 +58,47 @@ def post(self, request):
+    def get(self, request):
+        token = request.META.get('HTTP_AUTHORIZATION', '').split(' ')[1]
+        payload = jwt.decode(token,settings.SECRET_KEY)
+        RegistrationAPIView.send_link(self,email)
```
where is the email supposed to come from? you haven't declared the variable above.

ignatiusukwuoma commented:
You still have challenges properly rebasing your branch. I'm seeing changes made on @MaggieChege's PR. You will need to improve your knowledge and usage of Git and Github. I only want to see your changes showing up on this diff (Files Changed).

d-kahara committed:
refactor

-Create a utility.py file
-Refactor views

d-kahara commented:
> You still have challenges properly rebasing your branch. I'm seeing changes made on @MaggieChege's PR. You will need to improve your knowledge and usage of Git and Github. I only want to see your changes showing up on this diff (Files Changed).

Resolved

ignatiusukwuoma commented on authors/apps/authentication/utils.py:
```diff
@@ -0,0 +1,33 @@
+from django.template.loader import render_to_string
+
+# Method to Send emails
+def send_link(self, email):
```
This is no longer a method, it's a function. So it doesn't need to take `self` as an argument as self is an instance of a Class

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -4,6 +4,7 @@
 from django.conf import settings
 from django.template.loader import render_to_string
 from django.core.mail import send_mail
+from django.template.loader import render_to_string
```
This has been imported on line 5

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -32,11 +34,53 @@ def post(self, request):
+        token = request.META.get('HTTP_AUTHORIZATION', '').split(' ')[1]
+        payload = jwt.decode(token, settings.SECRET_KEY)
+        email = payload['email']
+        send_link(self, email)
```
No need for self

ignatiusukwuoma commented on authors/apps/authentication/views.py:
```diff
@@ -32,11 +34,53 @@ def post(self, request):
         serializer = self.serializer_class(data=user)
         serializer.is_valid(raise_exception=True)
+        email = serializer.validated_data.get('email', None)
+        send_link(self, email)
```
No need for self

ignatiusukwuoma commented:
We still have two templates doing same thing. Also, you will need to add more tests to increase coverage. Consider testing the views

ignatiusukwuoma commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,47 @@
+import json
+import jwt
+import os
```
Remove all unnecessary imports

d-kahara committed:
Refactor(tests):write more tests for email verification views

-Write more tests
-Decode token in user model to return string
-Resolve code convention issues

Georgeygigz commented on .env.sample:
```diff
@@ -9,3 +9,5 @@ export EMAIL_HOST_PASSWORD="email password"
 export DEFAULT_FROM_EMAIL="senders email"
 export SECRET_KEY="enter your secret key"
+export EMAIL_VERIFICATION_URL="http://127.0.0.1:8000/"
+EXPORT EXPIRED_LINK="http://127.0.0.1:8000/api/users/verify/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImJldnN0b3BAZ21haWwuY29tIiwiaWF0IjoxNTQ4MDYwMjIyLCJleHAiOjE1NDgwNjA1MjJ9.LpLr6O6zjCxtmJuURf15YGL5tkkLuCDjhj966fog6po/
```
export should be in lower case

kwameasiago commented on authors/apps/authentication/templates/email-verify.html:
```diff
@@ -0,0 +1,35 @@
+
+</body>
+
+</html>
```
Add a line at the end of the file

andrewhingah commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,65 @@
+        self.assertEqual(mail.outbox[0].subject, 'Authors Haven Verification Link')
+    
+    def test_expired_verification_link(self):
+        print(self.verify_url)
```
Don't push the print statement.

ignatiusukwuoma commented on authors/apps/authentication/models.py:
```diff
@@ -126,5 +128,5 @@ def token(self):
         }
-        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256')
-        return token.decode()
+        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256').decode('utf-8')
```
Probably no need to change this anymore

d-kahara replied:
I had to put in that part for tests to pass

ignatiusukwuoma commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,64 @@
+        self.client = APIClient()
+        self.signup_url = reverse('authentication:auth-register')
+        self.login_url = reverse('authentication:auth-login')
+        self.verify_url = os.getenv('EXPIRED_LINK')
```
If you only need EXPIRED_LINK for testing, then no need to put in the environment. Just have it in the test.

ignatiusukwuoma replied:
The instance variable name should also reflect that link is expired.

d-kahara replied:
I have to put it there since the environment changes when the app is deployed

ignatiusukwuoma commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_resend_verification_link(self):
+        register = self.client.post(self.signup_url,
```
Make your test DRY. make this a method and just call self.register() to register in your tests... if more than one test needs to register

ignatiusukwuoma commented on authors/apps/authentication/tests/test_email_verification.py:
```diff
@@ -0,0 +1,64 @@
+                                    format='json')
+        self.assertEqual(register.status_code, status.HTTP_201_CREATED)
+        token = register.data['token']
+        response = self.client.get(self.verify_url, HTTP_AUTHORIZATION='token {}'.format(token))
```
I got confused why this was a bad request until i checked the value of`self.verify_url`. The name should reflect the nature of the url.

ignatiusukwuoma commented on .env.sample:
```diff
@@ -9,3 +9,5 @@ export EMAIL_HOST_PASSWORD="email password"
 export DEFAULT_FROM_EMAIL="senders email"
 export SECRET_KEY="enter your secret key"
+export EMAIL_VERIFICATION_URL="http://127.0.0.1:8000/"
+export EXPIRED_LINK="http://127.0.0.1:8000/api/users/verify/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImJldnN0b3BAZ21haWwuY29tIiwiaWF0IjoxNTQ4MDYwMjIyLCJleHAiOjE1NDgwNjA1MjJ9.LpLr6O6zjCxtmJuURf15YGL5tkkLuCDjhj966fog6po/
```
Can we remove this from this file if it's no longer being used?

ignatiusukwuoma commented on authors/settings.py:
```diff
@@ -26,6 +26,14 @@
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_HOST_USER = os.getenv('EMAIL_HOST_USER')
+EMAIL_HOST_PASSWORD = os.getenv('EMAIL_HOST_PASSWORD')
+EMAIL_PORT = 587
```
These config have been added to the master already... at the base of this file
